### PR TITLE
Minimise cross track error

### PR DESCRIPTION
**Pull request by lthall**
This PR addresses a problem where a strong wind can overwhelm the aircrafts ability to hold the track. The problem is the aircraft does not divert a large enough proportion of the available lean angle away from forward speed to hold the track.
This PR results in the aircraft using 100% of the available lean angle to say on the track to the point the aircraft will stop moving along the track.
This PR should not impact normal operation.

The key changes are:

- providing access to accel limiting that prioritises cross track acceleration,
- limit the final acceleration of the position controller to prioritise acceleration in the cross track direction,
- allow the 2D PID loop I term to change direction, but not grow, when limited.

Master: 50m cross track error and growing:
MasterMission4

PR: 2.5m and stable:
PRMission4

Master: Square in wind:
MasterMission1

PR: Square in wind:
PRMission1

### CODE REVIEW

File: libraries/AC_PID/AC_PID_2D.cpp [hunk 1]
```diff
@@ -133,16 +133,17 @@ void AC_PID_2D::update_i(const Vector2f &limit)
     _pid_info_y.limit = false;
 
     Vector2f delta_integrator = (_error * _ki) * _dt;
+    float integrator_length = _integrator.length();
+    _integrator += delta_integrator;
     if (!is_zero(limit.length_squared())) {
```
Comment (IamPete1) [inline libraries/AC_PID/AC_PID_2D.cpp:138]:
```suggestion
    if (!is_zero(integrator_length)) {
```
small efficiency improvement.

File: libraries/AC_AttitudeControl/AC_PosControl.cpp [hunk 2]
```diff
@@ -639,11 +639,13 @@ void AC_PosControl::update_xy_controller()
...
 
     // limit acceleration using maximum lean angles
-    _limit_vector.xy().zero();
     float angle_max = MIN(_attitude_control.get_althold_lean_angle_max_cd(), get_lean_angle_max_cd());
     float accel_max = GRAVITY_MSS * 100.0f * tanf(ToRad(angle_max * 0.01f));
-    if (_accel_target.limit_length_xy(accel_max)) {
-        _limit_vector.xy() = _accel_target.xy();
+    // Define the limit vector before we constrain _accel_target 
+    _limit_vector.xy() = _accel_target.xy();
+    if ( !limit_accel_xy(_vel_desired.xy(), _accel_target.xy(), accel_max) ) {
```
Comment (rmackay9) [inline libraries/AC_AttitudeControl/AC_PosControl.cpp:646]: nitpick: we can remove the little space after the starting bracket

File: libraries/AP_Math/control.cpp [hunk 3]
```diff
@@ -324,6 +324,38 @@ void shape_pos_vel_accel_xy(const Vector2p& pos_input, const Vector2f& vel_input
     shape_vel_accel_xy(vel_target, accel_input, vel, accel, accel_max, jerk_max, dt, limit_total_accel);
 }
 
+// limit_accel_xy limits the acceleration to prioritise acceleration perpendicular to the provided velocity vector.
+bool limit_accel_xy(const Vector2f& vel, Vector2f& accel, float accel_max)
```
Comment (rmackay9) [inline libraries/AP_Math/control.cpp:328]: I think we can make the "accel" argument, "const Vector2f& accel"

File: libraries/AP_Math/control.cpp [hunk 4]
```diff
@@ -324,6 +324,38 @@ void shape_pos_vel_accel_xy(const Vector2p& pos_input, const Vector2f& vel_input
...
+{
+    // check accel_max is defined
+    if ( !is_positive(accel_max) ) {
```
Comment (rmackay9) [inline libraries/AP_Math/control.cpp:331]: nitpick: we can remove the spaces within the brackets.

File: libraries/AC_PID/AC_PID_2D.cpp [hunk 5]
```diff
@@ -133,16 +133,17 @@ void AC_PID_2D::update_i(const Vector2f &limit)
     _pid_info_y.limit = false;
 
     Vector2f delta_integrator = (_error * _ki) * _dt;
+    float integrator_length = _integrator.length();
```
Comment (tridge) [inline libraries/AC_PID/AC_PID_2D.cpp:136]: could be inside the if() ?

File: libraries/AP_Math/control.cpp [hunk 6]
```diff
@@ -324,6 +324,38 @@ void shape_pos_vel_accel_xy(const Vector2p& pos_input, const Vector2f& vel_input
...
+            // We do not have a direction of travel so do a simple vector length limit
+            accel.limit_length(accel_max);
+        } else {
+            // calculate acceleration in the direction of and perpendicular to the velocity input
+            const Vector2f vel_input_unit = vel.normalized();
+            // acceleration in the direction of travel
+            float accel_dir = vel_input_unit * accel;
+            // cross track acceleration
+            Vector2f accel_cross =  accel - (vel_input_unit * accel_dir);
+            if ( accel_cross.limit_length(accel_max) ) {
```
Comment (rmackay9) [inline libraries/AP_Math/control.cpp:346]: nitpick: these white spaces within the brackets can go..

File: libraries/AP_Math/control.h [hunk 7]
```diff
@@ -110,6 +110,9 @@ void shape_pos_vel_accel_xy(const Vector2p& pos_input, const Vector2f& vel_input
                             float vel_max, float accel_max,
                             float jerk_max, float dt, bool limit_total_accel);
 
+// limit_accel_xy limits the acceleration to prioritise acceleration perpendicular to the provided velocity vector.
+bool limit_accel_xy(const Vector2f& vel, Vector2f& accel, float accel_max);
```
Comment (rmackay9) [inline libraries/AP_Math/control.h:114]: I think the function itself can be made "const" because no internal state changes.. maybe?

File: libraries/AC_PID/AC_PID_2D.cpp (hunk 1 above)
Comment (lthall) [reply libraries/AC_PID/AC_PID_2D.cpp:138]: Agreed but I will move that down to remove the sqrt unless we need it.

File: libraries/AC_PID/AC_PID_2D.cpp (hunk 5 above)
Comment (lthall) [reply libraries/AC_PID/AC_PID_2D.cpp:136]: Done

File: libraries/AP_Math/control.cpp (hunk 4 above)
Comment (lthall) [reply libraries/AP_Math/control.cpp:331]: Done

File: libraries/AC_AttitudeControl/AC_PosControl.cpp (hunk 2 above)
Comment (lthall) [reply libraries/AC_AttitudeControl/AC_PosControl.cpp:646]: Done

File: libraries/AP_Math/control.cpp (hunk 6 above)
Comment (lthall) [reply libraries/AP_Math/control.cpp:346]: Done

**Comment by lthall** [comment]
All checks passed on my branch.

File: libraries/AP_Math/control.cpp [hunk 8]
```diff
@@ -324,6 +324,38 @@ void shape_pos_vel_accel_xy(const Vector2p& pos_input, const Vector2f& vel_input
...
+// limit_accel_xy limits the acceleration to prioritise acceleration perpendicular to the provided velocity vector.
+bool limit_accel_xy(const Vector2f& vel, Vector2f& accel, float accel_max)
+{
+    // check accel_max is defined
+    if (!is_positive(accel_max)) {
+        return false;
+    }
+    // limit acceleration to accel_max while prioritizing cross track acceleration
+    if (accel.length_squared() > sq(accel_max)) {
+        if (vel.is_zero()) {
```
Comment (hendjoshsr71) [inline libraries/AP_Math/control.cpp:336]: This is likely just being pedantic and I'm not sure on this one but `is.zero()` only compares each element. But we really  divide by the `length()`.

Should this be the following due to the `vec.normalize()`?

```suggestion
        if (is_zero(vel.length())) {
```
Or even more stringent ... 
```suggestion
        if (is_zero(vel.length_squared())) {
```

One other thought about `void Vector3F.normalize()`  
I'd be tempted to make a new `bool Vector3F.nomalized()` that does this division by zero check for us...

Looking through the code in some places we condition it upon `is_zero()`, `is_zero(vec.length())`, or `is_zero(vec.length())`

Basically some numerical consistency might be nice is our code. Thoughts?

Comment (lthall) [reply libraries/AP_Math/control.cpp:336]: I would say this should always be:
XXX.is_zero()
And that function should be the written to ensure that we can safely normalise the vector without getting a divide by zero.

So if the best option is to use the sum of the squares of each element to ensure that then we should change .iszero() to do that.

That is what I would think is the best way to do this.

Comment (lthall) [reply libraries/AP_Math/control.cpp:336]: I seem to recall a funny divide by zero that happened because while the vector was not zero (but very small), after converting it to a length it became zero. Maybe I am recalling incorrectly.

Comment (hendjoshsr71) [reply libraries/AP_Math/control.cpp:336]: Clearly, none of this holds back this PR. Just my numeric side likes consistency. 

@peterbarker thoughts on the this?

Changing `vec.is_zero()` is a far wider change and not something I would touch.  far more use cases and they are not consistent.

I'd be proposing every time we call `vec.normalize()` it would look like this because we do this anyways just with inconsistent methods using various metrics.
```cpp
if (!vec.normalize()) {
    /// vec length was zero
   // do stuff
}
```

instead of 
```cpp
if (!vec.is_zero() || ) {
    // vec length was zero
    // do stuff

} else {
    vec.normalize()
}
```

Comment (rmackay9) [reply libraries/AP_Math/control.cpp:336]: I'm not sure we could change .is_zero() because we don't really know how the vector is going to be used later.  even just multiplying the vector by 0.5 could be enough to change it from !is_zero() to is_zero().

Comment (hendjoshsr71) [reply libraries/AP_Math/control.cpp:336]: > I seem to recall a funny divide by zero that happened because while the vector was not zero (but very small), after converting it to a length it became zero. Maybe I am recalling incorrectly.

There was a case of length_squared check in one place but not in another for scurves. Which is why i'm a little keen on being consistent.

Comment (lthall) [reply libraries/AP_Math/control.cpp:336]: Using the equivalent of is_zero(lenght_squared)  instead of just checking each element is zero would just be a slightly more conservative test so I can't see that being a problem.

I would be happy to use .is_zero() or have .normalize() return a bool. Fixing .is_zero() to be a little more conservative using is_zero(lenght_squared) is probably less likely to trip up others in the future. I can see the attractiveness of if (vec.normalize()).

These are decisions I wait to get guidance from Tridge, Randy and Peter.

File: libraries/AP_Math/control.cpp [hunk 9]
```diff
@@ -324,6 +324,38 @@ void shape_pos_vel_accel_xy(const Vector2p& pos_input, const Vector2f& vel_input
     shape_vel_accel_xy(vel_target, accel_input, vel, accel, accel_max, jerk_max, dt, limit_total_accel);
 }
 
+// limit_accel_xy limits the acceleration to prioritise acceleration perpendicular to the provided velocity vector.
```
Comment (rmackay9) [inline libraries/AP_Math/control.cpp:327]: It would be good to add a comment saying when it returns true vs false.  I guess it is true if the passed in accel argument has been limited, false if it has been left unchanged.

**Review by rmackay9 — COMMENTED** [review]
nitpick: the commit message for the "AC_PID" change includes this description, "AC_PID: Squash"

File: libraries/AC_PID/AC_PID_2D.cpp [hunk 10]
```diff
@@ -133,16 +133,17 @@ void AC_PID_2D::update_i(const Vector2f &limit)
     _pid_info_y.limit = false;
 
     Vector2f delta_integrator = (_error * _ki) * _dt;
+    _integrator += delta_integrator;
     if (!is_zero(limit.length_squared())) {
-        // zero delta_vel if it will increase the velocity error
+        // do not let integrator increase in length if delta_integrator is in the direction of limit
         if (is_positive(delta_integrator * limit)) {
-            delta_integrator.zero();
+            float integrator_length = _integrator.length();
```
Comment (rmackay9) [inline libraries/AC_PID/AC_PID_2D.cpp:140]: It would definitely be good to test this change to AC_PID_2D with various values to ensure it is doing what is expected.  So for example, an AC_PID object could be temporarily created in the Copter 1hz loop and then various values passed in for the limit and then check that the resulting I-term is as expected.

These two lines seem a bit odd actually.. it seems that we take the length of the integrator and then immediately limit it to that length so this shouldn't have any effect right?  I think perhaps the idea was that we should take the length before the delta_integrator has been added?

Comment (rmackay9) [reply libraries/AC_PID/AC_PID_2D.cpp:140]: I guess there is no chance that we might suddenly shorten the integrator?  I think the intention is that we don't allow the integrator to grow in the direction of the limit.. but that's different than suddenly shortening it.

File: libraries/AC_PID/AC_PID_2D.cpp (hunk 1 above)
Comment (lthall) [reply libraries/AC_PID/AC_PID_2D.cpp:137]: This is actually the correct change. Sorry.

File: libraries/AC_PID/AC_PID_2D.cpp (hunk 10 above)
Comment (lthall) [reply libraries/AC_PID/AC_PID_2D.cpp:140]: Yes, this was a requested change to avoided calculating the length when we don't have to. We have to.

File: libraries/AC_PID/AC_PID_2D.cpp (hunk 5 above)
Comment (lthall) [reply libraries/AC_PID/AC_PID_2D.cpp]: No, we need to calculate this before we change that vector.

File: libraries/AC_PID/AC_PID_2D.cpp [hunk 11]
```diff
@@ -133,16 +133,17 @@ void AC_PID_2D::update_i(const Vector2f &limit)
     _pid_info_y.limit = false;
 
     Vector2f delta_integrator = (_error * _ki) * _dt;
-    if (!is_zero(limit.length_squared())) {
-        // zero delta_vel if it will increase the velocity error
+    float integrator_length = _integrator.length();
+    _integrator += delta_integrator;
+    if (!is_zero(integrator_length)) {
```
Comment (IamPete1) [inline libraries/AC_PID/AC_PID_2D.cpp:138]: I don't think we need this check. looks to me that`_integrator.limit_length` correctly deals with the 0 case.

Removing the check means we can remove the `integrator_length` float too.

It also means that we correctly set the pid_info limit flags even if there is 0 I term. Its a edge case but this could happen on the first loop after a I reset if you have a large I gain and small I max.

File: libraries/AC_PID/AC_PID_2D.cpp [hunk 12]
```diff
@@ -133,16 +133,17 @@ void AC_PID_2D::update_i(const Vector2f &limit)
...
+        // do not let integrator increase in length if delta_integrator is in the direction of limit
         if (is_positive(delta_integrator * limit)) {
-            delta_integrator.zero();
+            _integrator.limit_length(integrator_length);
```
Comment (IamPete1) [inline libraries/AC_PID/AC_PID_2D.cpp:141]:
```suggestion
                   if (is_positive(delta_integrator * limit) && _integrator.limit_length(integrator_length)) {
```
A change to check the return of `limit_length`

Comment (lthall) [reply libraries/AC_PID/AC_PID_2D.cpp:141]: I like that!

File: libraries/AC_PID/AC_PID_2D.cpp (hunk 11 above)
Comment (lthall) [reply libraries/AC_PID/AC_PID_2D.cpp]: This is done but I don't see how we can remove integrator_length.

File: libraries/AC_PID/AC_PID_2D.cpp (hunk 12 above)
Comment (lthall) [reply libraries/AC_PID/AC_PID_2D.cpp]: Done, Thanks.

File: libraries/AP_Math/control.cpp (hunk 8 above)
Comment (lthall) [reply libraries/AP_Math/control.cpp:341]: This is probably something worth discussing at the next dev meeting.

File: libraries/AC_PID/AC_PID_2D.cpp [hunk 13]
```diff
@@ -133,16 +133,14 @@ void AC_PID_2D::update_i(const Vector2f &limit)
...
-        // zero delta_vel if it will increase the velocity error
-        if (is_positive(delta_integrator * limit)) {
-            delta_integrator.zero();
-            _pid_info_x.limit = true;
-            _pid_info_y.limit = true;
-        }
+    float integrator_length = _integrator.length();
+    _integrator += delta_integrator;
+    // do not let integrator increase in length if delta_integrator is in the direction of limit
+    if (is_positive(delta_integrator * limit) && _integrator.limit_length(integrator_length)) {
```
Comment (IamPete1) [inline libraries/AC_PID/AC_PID_2D.cpp:139]: You can remove the float by moving it to a direct call in the if. This means that it will only be evaluated if `is_positive(delta_integrator * limit)` is true.  
```suggestion
    if (is_positive(delta_integrator * limit) && _integrator.limit_length(_integrator.length())) {
```

Comment (lthall) [reply libraries/AC_PID/AC_PID_2D.cpp:139]: No, we need to calculate the length before we add the delta_integrator so we can't do this.

Comment (IamPete1) [reply libraries/AC_PID/AC_PID_2D.cpp:139]: yep, your right. I missed that.

**Comment by rmackay9** [comment]
Merged, thanks!
